service/discovery.go

```diff
@@ -0,0 +1,75 @@
+package service
+
+import (
+	"fmt"
+	"net/url"
+	"time"
+
+	"github.com/hashicorp/consul/api"
+	"golang.org/x/sync/syncmap"
+)
+
+// DefaultCacheInvalidationTimeout for updating instances
+const (
+	DefaultCacheInvalidationTimeout = 30
+)
+
+// Services for discovery service
+type Services struct {
+	ConsulClient             *api.Client
+	Instances                *syncmap.Map
+	CacheInvalidationTimeout int64
+}
+
+// GetEndpoint by service name
+func (s *Services) GetEndpoint(serviceName string) (url *url.URL, err error) {
+	resolver := s.UpdateInstences(serviceName)
+	if resolver != nil && len(resolver.endpoints) > 0 {
+		url = resolver.getHealthyInstanceEndpoint()
+	} else {
+		err = fmt.Errorf("no registered or healtly instances for service: %s", serviceName)
+	}
+	return
+}
+
+// UpdateInstences get service instances from service discovery
+func (s *Services) UpdateInstences(serviceName string) (resolver *Resolver) {
+	value, instancesExists := s.Instances.Load(serviceName)
+	if instancesExists {
+		resolver = value.(*Resolver)
+	} else {
+		s.Instances = new(syncmap.Map)
```

Comment:
Wait, if I understand correctly, then assuming that `s.Instances` has entries for services `B` and `C` and a resolver for service `A` doesn't exist, then we flush the entries for services `B` and `C`.  
Right?